crates/wasi-common/winx/src/file.rs

```diff
@@ -415,3 +388,103 @@ pub fn change_time(file: &File) -> io::Result<i64> {
 
     Ok(tm)
 }
+
+pub fn query_access_information(handle: RawHandle) -> Result<AccessMode> {
+    let mut io_status_block = IO_STATUS_BLOCK::default();
+    let mut info = FILE_ACCESS_INFORMATION::default();
+
+    unsafe {
+        let status = NtQueryInformationFile(
+            handle,
+            &mut io_status_block,
+            &mut info as *mut _ as *mut c_void,
+            std::mem::size_of::<FILE_ACCESS_INFORMATION>() as u32,
+            FILE_INFORMATION_CLASS::FileAccessInformation,
+        );
+
+        if status != ntstatus::STATUS_SUCCESS {
+            return Err(winerror::WinError::from_u32(RtlNtStatusToDosError(status)));
+        }
+    }
+
+    Ok(AccessMode::from_bits_truncate(info.AccessFlags))
+}
+
+pub fn query_mode_information(handle: RawHandle) -> Result<FileModeInformation> {
+    let mut io_status_block = IO_STATUS_BLOCK::default();
+    let mut info = FILE_MODE_INFORMATION::default();
+
+    unsafe {
+        let status = NtQueryInformationFile(
+            handle,
+            &mut io_status_block,
+            &mut info as *mut _ as *mut c_void,
+            std::mem::size_of::<FILE_MODE_INFORMATION>() as u32,
+            FILE_INFORMATION_CLASS::FileModeInformation,
+        );
+
+        if status != ntstatus::STATUS_SUCCESS {
+            return Err(winerror::WinError::from_u32(RtlNtStatusToDosError(status)));
+        }
+    }
+
+    Ok(FileModeInformation::from_bits_truncate(info.Mode))
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_snake_case)]
+enum FILE_INFORMATION_CLASS {
```

Comment:
Does the enum name need to be all caps? Or could we use a Rust-idiomatic way, i.e., `enum FileInformationClass`?

Reply:
Like the `winapi` crate, I tend to like the Win32 naming convention to call out that these are interop types, but I am definitely open to renaming them if we all feel that's best.

Reply:
Yeah, I'm kinda on the fence with this one. I am personally not a fan of all caps for non-consts, but then again, this is indeed the Win32 naming convention. Hmm, I hope this is not an overkill, but perhaps we could compromise here and go with the Rust convention, i.e., keep the naming in the spirit of Win32, but move the code to a separate internal module in the spirit of [c.rs] in Rust's libstd? What do you think? We could then get rid of `allow(non_snake_case)` in favour of a single `allow(nonstandard_style)`.

[c.rs]: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/c.rs

Reply:
I think moving this out into a separate file is a good idea.  I'll do that.

Reply:
It's a good idea to move it into a separate crate but only if it's not possible to add them to winapi.

Reply:
I think a separate crate might be warranted once we have a few more ntdll APIs needed; for a single ntdll export it seems like unnecessary tooling overhead.  I don't hold a strong opinion on this, though.

Reply:
I've moved the `ntdll` related types and functions out to a separate module named `ntdll` for now.  I think we can revisit the need for another crate if this file ever grows (hopefully it won't).

Reply:
I agree with @peterhuene here. I'd wait until we start using more and more from `ntdll` to warrant its inclusion in `winapi`. Until then, I'd keep it local for our internal use only.